src/util/string.cpp

```diff
@@ -858,10 +903,37 @@ void translate_all(const std::wstring &s, size_t &i,
 		} else if (parts[0] == L"T") {
 			// Beginning of translated string.
 			std::wstring textdomain;
+			bool use_plural = false;
+			unsigned long int number = 0;
 			if (parts.size() > 1)
 				textdomain = parts[1];
+			if (parts.size() > 2 && parts[2] != L"") {
+				// parts[2] should contain a number used for selecting
+				// the plural form.
+				// However, we can't blindly cast it to an unsigned long int,
+				// as it might be too large for that.
+				//
+				// We follow the advice of gettext and reduce integers larger than 1000000
+				// to something in the range [1000000, 2000000), with the same last 6 digits.
+				//
+				// https://www.gnu.org/software/gettext/manual/html_node/Plural-forms.html
+				unsigned long int max = 1000000;
+
+				use_plural = true;
+				number = 0;
+				for (char c : parts[2]) {
+					if (L'0' <= c && c <= L'9') {
+						number = (10 * number + (unsigned long int)(c - L'0'));
+						if (number >= 2 * max) number = (number % max) + max;
+					} else {
+						// Invalid number
+						use_plural = false;
+						break;
+					}
+				}
+			}
```

Comment:
Shouldn't this already be done in Lua? Seems like it could just convert the string to a number.

Reply:
Unfortunately not: Lua has to convert the number to a string to encode it inside the translation string. However, it is possible to create invalid translation strings without using `minetest.translate`, and these must not crash the code.

Reply:
Oh my bad I thought the limit was just to avoid incorrect translations.

Reply:
Ah, you meant the truncation to 6 digits should already be done by Lua? In that case yes, but I prefer to be conservative about the inputs received from the server. However, I can remove it if you prefer.

Reply:
No I suppose it's good to make sure, could you make `max` a `constexpr`?